[FIRRTL] Chisel generates FIRRTL which fails new Cat signedness check

I was trying the new `firtool-1.123.0` release, and am running into an issue with the FIRRTL generated by RocketChip, and its interaction with a new check added in https://github.com/llvm/circt/pull/8557. The code in question is here: https://github.com/llvm/circt/blob/211c3faba6aad6e2d57efe7a76d78678ab426184/lib/Dialect/FIRRTL/FIRRTLOps.cpp#L5099-L5116

I'm getting this error message:

```
generators/rocket-chip/src/main/scala/rocket/RocketCore.scala:1385:8: error: all operands must have same signedness
generators/rocket-chip/src/main/scala/rocket/RocketCore.scala:1385:8: error: 'firrtl.cat' op failed to infer returned types
generators/rocket-chip/src/main/scala/rocket/RocketCore.scala:1385:8: note: see current operation: %1737 = "firrtl.cat"(%1697, %1702, %1709, %1719, %1722, %1731, %1736) {name = "_ex_imm_T"} : (!firrtl.sint<1>, !firrtl.sint<11>, !firrtl.sint<8>, !firrtl.sint<1>, !firrtl.uint<6>, !firrtl.uint<4>, !firrtl.uint<1>) -> !firrtl.uint<32>
```

The Chisel code which causes this issue can be found here: https://github.com/chipsalliance/rocket-chip/blob/c357a0f32ecec9ee046ca1d6378c21ec7fec179b/src/main/scala/rocket/RocketCore.scala#L1369-L1387

I'm on Chisel 7.0.0-RC1+37-f8a2dec3-SNAPSHOT.

Is this a situation where Chisel is generating illegal FIRRTL? Or a situation where the new variadic form of `firrtl.cat` is interacting with a weird edge-case in FIRRTL? [The FIRRTL Spec says](https://github.com/chipsalliance/firrtl-spec/blob/b9ca6a6d1ba6946ed71663facaf0177f27088107/spec.md#L3964-L3971) that the `cat` function has two legal signatures: `(UInt, UInt) -> SInt` or `(SInt, SInt) -> UInt`. I can imagine that, since each two-operand `cat` generates a `UInt`, so long as you create the appropriate tree of `cat` operators, you can get a legal `cat` result from many arguments which have mixed signedness. For example: `cat(cat(SInt, SInt), UInt) -> cat(UInt, UInt) -> UInt`. But if you fold this into a 3-input `cat`, you get `cat(SInt, SInt, UInt)` which fails this check.

Any idea how to resolve this?

--- a/lib/Dialect/FIRRTL/FIRRTLFolds.cpp
+++ b/lib/Dialect/FIRRTL/FIRRTLFolds.cpp
@@ -1404,24 +1404,42 @@ public:
         worklist.push_back(operand);
     };
     pushOperands(cat);
+    bool existsSigned = false, existsUnsigned = false;
     while (!worklist.empty()) {
       auto value = worklist.pop_back_val();
       auto catOp = value.getDefiningOp<CatPrimOp>();
       if (!catOp) {
         operands.push_back(value);
+        (type_isa<UIntType>(value.getType()) ? existsUnsigned : existsSigned) =
+            true;
         continue;
       }
 
       pushOperands(catOp);
     }
 
+    // Helper function to cast signed values to unsigned. CatPrimOp converts
+    // signed values to unsigned so we need to do it explicitly here.
+    auto castToUIntIfSigned = [&](Value value) -> Value {
+      if (type_isa<UIntType>(value.getType()))
+        return value;
+      return rewriter.create<AsUIntPrimOp>(value.getLoc(), value);
+    };
+
+    assert(operands.size() >= 1 && "zero width cast must be rejected");
+
     if (operands.size() == 1) {
-      rewriter.replaceOp(op, operands[0]);
+      rewriter.replaceOp(op, castToUIntIfSigned(operands[0]));
       return success();
     }
 
     if (operands.size() == cat->getNumOperands())
       return failure();
+
+    // If types are mixed, cast all operands to unsigned.
+    if (existsSigned && existsUnsigned)
+      for (auto &operand : operands)
+        operand = castToUIntIfSigned(operand);
 
     replaceOpWithNewOpAndCopyName<CatPrimOp>(rewriter, op, cat.getType(),
                                              operands);

--- a/lib/Dialect/FIRRTL/FIRRTLFolds.cpp
+++ b/lib/Dialect/FIRRTL/FIRRTLFolds.cpp
@@ -1404,14 +1404,13 @@ public:
         worklist.push_back(operand);
     };
     pushOperands(cat);
-    bool existsSigned = false, existsUnsigned = false;
+    bool hasSigned = false, hasUnsigned = false;
     while (!worklist.empty()) {
       auto value = worklist.pop_back_val();
       auto catOp = value.getDefiningOp<CatPrimOp>();
       if (!catOp) {
         operands.push_back(value);
-        (type_isa<UIntType>(value.getType()) ? existsUnsigned : existsSigned) =
-            true;
+        (type_isa<UIntType>(value.getType()) ? hasUnsigned : hasSigned) = true;
         continue;
       }
 
@@ -1437,7 +1436,7 @@ public:
       return failure();
 
     // If types are mixed, cast all operands to unsigned.
-    if (existsSigned && existsUnsigned)
+    if (hasSigned && hasUnsigned)
       for (auto &operand : operands)
         operand = castToUIntIfSigned(operand);
 

--- a/test/Dialect/FIRRTL/canonicalization.mlir
+++ b/test/Dialect/FIRRTL/canonicalization.mlir
@@ -1066,6 +1066,25 @@ firrtl.module @FlattenCatEdgeCases(in %a: !firrtl.uint<4>, out %out1: !firrtl.ui
   firrtl.matchingconnect %out2, %zero_cat2 : !firrtl.uint<8>
 }
 
+// Issue 8618
+// CHECK-LABEL: firrtl.module @FlattenCatDifferentSigns
+firrtl.module @FlattenCatDifferentSigns(in %a: !firrtl.uint<1>, in %b: !firrtl.sint<1>, in %c: !firrtl.sint<1>, out %d: !firrtl.uint<3>, out %e: !firrtl.uint<4>) {
+  // CHECK-NEXT: %[[AS_UINT1:.+]] = firrtl.asUInt %b : (!firrtl.sint<1>) -> !firrtl.uint<1>
+  // CHECK-NEXT: %[[AS_UINT2:.+]] = firrtl.asUInt %b : (!firrtl.sint<1>) -> !firrtl.uint<1>
+  // CHECK-NEXT: %[[CAT1:.+]] = firrtl.cat %a, %[[AS_UINT1]], %[[AS_UINT2]] : (!firrtl.uint<1>, !firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<3>
+  // CHECK-NEXT: firrtl.matchingconnect %d, %[[CAT1]]
+  %0 = firrtl.cat %b, %b : (!firrtl.sint<1>, !firrtl.sint<1>) -> !firrtl.uint<2>
+  %1 = firrtl.cat %a, %0 : (!firrtl.uint<1>, !firrtl.uint<2>) -> !firrtl.uint<3>
+  firrtl.matchingconnect %d, %1 : !firrtl.uint<3>
+
+  // CHECK-NEXT: %[[CAT2:.+]] = firrtl.cat %b, %c, %c, %b : (!firrtl.sint<1>, !firrtl.sint<1>, !firrtl.sint<1>, !firrtl.sint<1>) -> !firrtl.uint<4>
+  // CHECK-NEXT: firrtl.matchingconnect %e, %[[CAT2]]
+  %2 = firrtl.cat %b, %c : (!firrtl.sint<1>, !firrtl.sint<1>) -> !firrtl.uint<2>
+  %3 = firrtl.cat %c, %b : (!firrtl.sint<1>, !firrtl.sint<1>) -> !firrtl.uint<2>
+  %4 = firrtl.cat %2, %3 : (!firrtl.uint<2>, !firrtl.uint<2>) -> !firrtl.uint<4>
+  firrtl.matchingconnect %e, %4 : !firrtl.uint<4>
+}
+
 // CHECK-LABEL: firrtl.module @CatOfConstant
 firrtl.module @CatOfConstant(in %a: !firrtl.uint<4>, out %out1: !firrtl.uint<12>,
                             out %out2: !firrtl.uint<10>) {

